### PR TITLE
AST/SILGen: Use @_alwaysEmitIntoClient diagnostic helper in unavailable code

### PR DESCRIPTION
**Pull request by tshortli**
The `_diagnoseUnavailableCodeReached()` function was introduced in the Swift 5.9 standard library and employs `@backDeployed` to support compilation of binaries that target OS releases aligned with earlier Swift releases. Unfortunately, though, this back-deployment strategy doesn't work well for some unusual build environments. Specifically, in some configurations code may be built with a compiler from a recent Swift toolchain and then linked against the dylibs in an older toolchain. When linking against the older dylibs, the `_diagnoseUnavailableCodeReached()` function does not exist but the `@backDeployed` thunks emitted into the binary reference that function and therefore linking fails.

The idea of building with one toolchain and then linking to the dylibs in a different, older toolchain is extremely dubious. However, it exists and for now we need to support it. This PR introduces an alternative `_diagnoseUnavailableCodeReached()` function that is annotated with `@_alwaysEmitIntoClient`. Calls to the AEIC variant are now emitted by the compiler when the deployment target is before Swift 5.9.

Once these unusual build environments upgrade and start linking against a Swift 5.9 toolchain or later we can revert all of this.

Resolves rdar://119046537

### CODE REVIEW

**Comment by tshortli** [comment]
@swift-ci please test

File: stdlib/public/core/AssertCommon.swift [hunk 1]
```diff
@@ -329,6 +329,18 @@ internal func _diagnoseUnexpectedEnumCase<SwitchedValue>(
...
+}
+
+// FIXME: Remove this with rdar://119892482
+/// An `@_alwaysEmitIntoClient` variant of `_diagnoseUnavailableCodeReached()`.
+/// This is temporarily needed by the compiler to reference from back deployed
+/// clients.
+@_alwaysEmitIntoClient
+@inline(never)
+@_semantics("unavailable_code_reached")
+@usableFromInline // COMPILER_INTRINSIC
```
Comment (Azoy) [inline stdlib/public/core/AssertCommon.swift:342]: Don't think this is needed if it's already aeic

Comment (tshortli) [reply stdlib/public/core/AssertCommon.swift:342]: Yeah you're right. I'll remove it once CI has completed an initial run.

**Comment by tshortli** [comment]
@swift-ci please test

**Pull request merged**